src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs

```diff
@@ -1247,9 +1261,32 @@ private void WriteMetadata(string key, string value)
                 }
             }
 
-            private void WriteCopyLocalMetadata(LockFileTargetLibrary package, string assetsFileName, string destinationSubDirectory = null)
+            private bool WriteCopyLocalMetadataIfNeeded(LockFileTargetLibrary package, string assetsFileName, string destinationSubDirectory = null)
             {
-                WriteMetadata(MetadataKeys.CopyLocal, "true");
+                bool shouldCopyLocal = true;
+                if (_copyLocalPackageExclusions != null && _copyLocalPackageExclusions.Contains(package.Name))
+                {
+                    shouldCopyLocal = false;
+                }
+                bool shouldIncludeInPublish = shouldCopyLocal;
+                if (shouldIncludeInPublish && _publishPackageExclusions != null && _publishPackageExclusions.Contains(package.Name))
+                {
+                    shouldIncludeInPublish = false;
+                }
+
+                if (shouldCopyLocal == false && shouldIncludeInPublish == false)
+                {
+                    return false;
+                }
+
+                if (shouldCopyLocal)
+                {
+                    WriteMetadata(MetadataKeys.CopyLocal, "true");
+                    if (!shouldIncludeInPublish)
+                    {
+                        WriteMetadata(MetadataKeys.CopyToPublishDirectory, "false");
```

Comment:
As we're now writing out additional metadata, do we need to rev the cache format version number?

Reply:
It's safe to leave the version unchanged when only adding metadata. The metadata count is recorded in the file:

https://github.com/dotnet/sdk/blob/5c959e93e6cdda30049b02c945712d2798edb14d/src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs#L259-L266